test/unit/offset.js

```diff
@@ -568,4 +568,142 @@ QUnit.test( "iframe scrollTop/Left (see gh-1945)", function( assert ) {
 	}
 } );
 
+(function() {
+	var POSITION_VALUES = [ "static", "relative", "absolute", "fixed" ];
+
+	supportjQuery.each( POSITION_VALUES, function( i, docPosition ) {
+		supportjQuery.each( POSITION_VALUES, function( i, bodyPosition ) {
+			supportjQuery.each( [ "static", "absolute" ], function( i, markerPosition ) {
+
+				testIframe( "coordinates relative to document" +
+							" (html{position:" + docPosition + "} body{position:" + bodyPosition + "} div{position:" + markerPosition + "})",
+						"offset/rel-doc.html", function( assert, $, iframe, doc ) {
+					assert.expect( 256 );
+
+					var docElem = doc.documentElement, // `<html>` element
+						bodyElem = doc.body,
+						marker1 = doc.getElementById( "marker1" ),
+						marker2 = doc.getElementById( "marker2" ),
+						$marker1 = $( marker1 ),
+						$marker2 = $( marker2 ),
+						affectProps, markerTopLeft,
+
+						// Bit as Flag of box-properties
+						FLAG_DOC_MARGIN = 1,
+						FLAG_DOC_BORDER = 2,
+						FLAG_DOC_PADDING = 4,
+						FLAG_BODY_MARGIN = 8,
+						FLAG_BODY_BORDER = 16,
+						FLAG_BODY_PADDING = 32,
```

Comment:
> Sorry, I have to learn English...
So, please check following:

No problem. And it's not all your misunderstanding; my suggestions are changing as I learn new details of browser behavior around this issue.

> I interpreted [your comment in that issue](https://github.com/jquery/jquery/issues/3080#issuecomment-215738621) as that the code should be tested with changing `margin` and `border`. Did I mistake again?

It was a partial mistake. I wasn't saying that margin and border need to _change_, just that test cases should cover both zero and nonzero values. Further, we already have tests for the no-margin no-border no-padding cases, and I have since concluded that the "only margin" and "only border" cases are unnecessary, so this PR can focus on situations in which **all** the box styles are nonzero.

> Return values from `.offset()` have to be copied to be checked by `assert.deepEqual()`… `supportjQuery.extend({}, $( "#static" ).offset())`

:+1:

> If `body` has `position:non-static`, the element is positioned relative to `body` without being affected by `position` of `<html>`. `body` was already positioned with `margin`, `border` and `padding` of `<html>`.
Therefore the code should be:

> ```js
var origin =
    bodyPosition !== "static" ? 10 + 20 + 40 + 80 + 160 : // `10 + 20 + 40` must be included
    docPosition !== "static" ? 10 + 20 : // relative to inside `<html>`
    0; // relative to document (i.e. outside `<html>`)
```

I definitely mis-defined `origin`, and I agree that your fix is correct.

> I think, this code is incorrect:

> ```js
if ( docPosition !== "static" ) {
    originAdjust += 15 - (docPosition === "relative" ? 0 : 10);
}
```

> If `<html>` has `position:non-static`, it is positioned with `margin` even if `potision` is other than `relative`.
Therefore the code should be:

> ```js
if ( docPosition !== "static" ) {
    originAdjust += 15;
}
```

Again, a correct fix for my mistake.

> The test code I wrote might do the almost same as your code, by removing code that changes `margin` and `border`.
But I didn't think that it should change `top` and `left` of `<html>` and `body`.
Therefore I think that I can't write PR. Maybe, I can't understand your request well by my very poor English.
Then, I hope that someone write PR.

I really think we're close. My complaint wan't about what you were testing, just how it was being tested. And as shown above, you clearly have the right insights.